Reuse path variable for bodyfile

eloo-abi opened the pull request:
Hi, 
this PR enables the usage of the variables defined in the imposters to be reused in the bodyFile path.
So we can use "dynamic" responses for different paths easily.
Fixes: #175

eloo-abi committed:
feat: replace variables in bodyFile path

eloo-abi committed:
build: add test as make goal

joanlopez commented:
Hey @eloo-abi! Thanks for your contribution 💟  However, before moving forward, I'd like to see some tests that show the expected behavior in different scenarios (e.g. no variables at all, a variable with a single value, a variable with multiple values, multiple variables, when a non-defined variable is used, etc).

eloo-abi commented:
Hi, 

> no variables at all
this is covered by the existing tests already as the implementation is not going to change anything here

> a variable with a single value
will be added
 
> a variable with multiple values
not sure what this is, can you give an example of a variable with multiple values? 

> multiple variables
will be added

> when a non-defined variable is used
it behaves like a wrong manually entered path as the implementation is not doing anything special, just adjust the path to search for a file

eloo-abi committed:
test: add tests for variables in body file

joanlopez commented:
> Hi,
> 
> > no variables at all
> 
> this is covered by the existing tests already as the implementation is not going to change anything here
> 
> > a variable with a single value
> 
> will be added
> 
> > a variable with multiple values
> 
> not sure what this is, can you give an example of a variable with multiple values?
> 
> > multiple variables
> 
> will be added
> 
> > when a non-defined variable is used
> 
> it behaves like a wrong manually entered path as the implementation is not doing anything special, just adjust the path to search for a file

Yes, please, ignore the case of multiple variables. Cover the rest, and we can go! 🚀

eloo-abi committed:
remove unused line

eloo-abi committed:
test: add tests for variables in body file

Merge in OSSC/killgrave from feature/GH-175-reuse-path-variable-for-bodyfile to feat/GH-175-reuse-path-variable-for-bodyfile

* commit '18e28e7f7c16c2a4fc670f2aae6b744a82c1be51':
  remove unused line
  test: add tests for variables in body file

eloo-abi commented:
tests added 

i was not sure where or how to add to i have added it similar to other `handler_tests`
i hope this makes sense

eloo-abi commented:
@joanlopez 
hi, any news near? 
if everything fits your needs maybe we can merge this?

thanks

joanlopez commented on internal/server/http/handler_test.go:
```diff
@@ -69,6 +71,59 @@ func TestImposterHandler(t *testing.T) {
+
+	var imposters []Imposter
+	err := json.Unmarshal(imposterBytes, &imposters)
+	assert.NoError(t, err)
```
```suggestion
	require.NoError(t, err)
```

The recommendation for error-related assertions is `require`.

joanlopez commented on internal/server/http/handler_test.go:
```diff
@@ -69,6 +71,59 @@ func TestImposterHandler(t *testing.T) {
+		{"valid imposter with id 1 and second variable 2 in path", imposters[1], "/gophers/1/2", responseId1Variable2, http.StatusOK},
+	}
+
+	for _, tt := range dataTest {
+		t.Run(tt.name, func(t *testing.T) {
```
```suggestion
	for _, tt := range dataTest {
	        tt := tt
		t.Run(tt.name, func(t *testing.T) {
```

For safety, until we upgrade minimum Go version from v1.21 to v1.22, where `loopvar` is fixed.

joanlopez commented on internal/server/http/handler_test.go:
```diff
@@ -69,6 +71,59 @@ func TestImposterHandler(t *testing.T) {
+			m.Handle(tt.imposter.Request.Endpoint, handler)
+			m.ServeHTTP(rec, req)
+
+			expectedBodyPathFile, _ := os.Open(tt.expectedBodyPath)
```
```suggestion
			expectedBodyPathFile, err := os.Open(tt.expectedBodyPath)
			require.NoError(t, err)
```

joanlopez commented on internal/server/http/test/testdata/imposters_variables/responses/gopher_1_2_response.json:
```diff
@@ -0,0 +1,11 @@
+{
```
Can we have different attributes (name, color and age) for each, please? I know it may sound nitpicky, but when things go wrong, this is the kind of things that makes the difference at time of finding the issue.

joanlopez commented on internal/server/http/handler_test.go:
```diff
@@ -69,6 +71,59 @@ func TestImposterHandler(t *testing.T) {
+		expectedBodyPath string
+		statusCode       int
+	}{
+		{"valid imposter with id 1 in path", imposters[0], "/gophers/1", responseId1, http.StatusOK},
```
Can you add a test case where there's a value defined in the `bodyFile` path but not in the `endpoint`, please? Just to reflect what's the expectation in such case.

eloo-abi replied:
for sure.. 
but this is the same behaviour as right now if the file could not be found 

"200" with empty body

joanlopez commented:
> @joanlopez hi, any news near? if everything fits your needs maybe we can merge this?
> 
> thanks

Hey @eloo-abi! Sorry for the delay, I've been quite busy during recent weeks.
I left some comments and suggestions, I think that, once fixed, it will be good enough to be merged, for me.

Thanks! 🙇🏻

eloo-abi committed:
fix: adjust test data and handle errors

eloo-abi committed:
Pull request #3: fix: adjust test data and handle errors

Merge in OSSC/killgrave from feature/GH-175-reuse-path-variable-for-bodyfile to feat/GH-175-reuse-path-variable-for-bodyfile

* commit 'c9dbb4b7afa8c138d678ba338d7f4be397297034':
  fix: adjust test data and handle errors

eloo-abi commented:
@joanlopez everything should be addressed :)

eloo-abi commented:
@joanlopez did you have already time to have a look?

joanlopez commented:
> @joanlopez did you have already time to have a look?

Been AFK for some time during September, but planning to review it during October for sure! 🙇🏻

joanlopez commented on internal/server/http/handler_test.go:
```diff
@@ -69,6 +71,66 @@ func TestImposterHandler(t *testing.T) {
+	responseId2 := "test/testdata/imposters_variables/responses/gopher_2_response.json"
+	responseId1Variable1 := "test/testdata/imposters_variables/responses/gopher_1_1_response.json"
+	responseId1Variable2 := "test/testdata/imposters_variables/responses/gopher_1_2_response.json"
+	responseId1WithoutVariable := "test/testdata/imposters_variables/responses/gopher_1_without_variable_response.json"
```
Nit; I think this could be refactored in a way so we use the file contents in the test (instead of the file path), so we don't need an empty file for the case where there's no data, but it's not critical/blocking for merging this PR.

joanlopez reviewed (dismissed):
Hey! I left a minor comment, in case you want to address it, but overall looks good! 👍🏻 
Thanks 🙇🏻

joanlopez commented:
> @joanlopez if its not a blocker please merge this PR as it takes already a very long time for such a small change

Yeah, sure! I'm waiting for @aperezg to give his 👍🏻, which I hope we get in the next few days, it's alright from my side! 🙇🏻

eloo-abi commented:
@joanlopez @aperezg any updates here?

eloo-abi committed:
Merge branch 'main' into feat/GH-175-reuse-path-variable-for-bodyfile

eloo-abi commented:
@joanlopez @aperezg any updates here?

eloo-abi commented:
any news here?

or is the project now officially dead?

eloo-abi commented:
@joanlopez @aperezg any updates here?

joanlopez commented:
> @joanlopez @aperezg any updates here?

Hi @eloo-abi,

We're willing to continue working on this project, but the circumstances aren't easy right now, so we cannot make any promise in the short term (2-4 weeks). We might start resuming activity mid term, or eventually archive the project if completely unable to resume the maintenance.

Will let you know once we move on. Thanks for reaching! 🙇🏻
